File: hw7/hw6_midtermexam.cpp
// Copyright 2023 Yuqi Jin yuqijin8@bu.edu
// Copyright 2023 Shivam Goyal sgoyal15@bu.edu

#include <iostream>
#include <vector>
#include <map>
#include <stdexcept>
#include <fstream>
#include <string>
#include <iomanip>
#include "json.hpp"
using nlohmann::json;
using std::cout;

// part 1
std::vector<std::string> translate(const std::vector<std::string> msg, const std::map<std::string, std::string> convert) {
    std::vector<std::string> ans_list;

    for (const auto str: msg) {
        try {
            if (convert.find(str) != convert.end()) {
                ans_list.push_back(convert.at(str));
            } else {
                ans_list.push_back(str);
            }
        } catch (...) {
            if (convert.empty()) {
                for (const auto& element : msg) {
                    ans_list.push_back(element);
                }
            }
        }
    }
    return ans_list;
}
// int main() {
//     const std::vector<std::string> message = {"I", "am", "Paul"}; // Use string literals instead of integers
//     const std::map<std::string, std::string> convert = {};

//     std::vector<std::string> translated = translate(message, convert);

//     for (const auto& word : translated) {
//         std::cout << word << " ";
//     }
//     std::cout << std::endl;

//     return 0;
// }

// Part 2
class SmartBulb {
public:
    SmartBulb(const std::array<int, 4>& ip, int port) {
        if (!isValidIP(ip)) {
            throw std::invalid_argument("IPv4 address must be a 4-tuple of 8-bit integers");
        }
        if (!isValidPort(port)) {
            throw std::invalid_argument("Port number must be an int16 value [0, 65535]");
        }

        this->ip = ip;
        this->port = port;
        this->lightState = "undeclared";
        this->brightnessLevel = 0.0;
    }
    // turn on
    void on() {
        lightState = "on";
    }
    // turn off
    void off() {
        lightState = "off";
    }

    void dim(double brightness) {
        if (!isValidBrightness(brightness)) {
            throw std::invalid_argument("Brightness value must be [0,1]");
        }
        lightState = "dim";
        brightnessLevel = brightness;
    }

    double brightness() const {
        if (lightState == "on") {
            return 1.0;
        } else if (lightState == "dim") {
            return brightnessLevel;
        } else {
            return 0.0;
        }
    }

    operator bool() const {
        if (lightState == "on" || lightState == "dim") {
            return true;
        } else if (lightState == "off") {
            return false;
        } else {
            throw std::logic_error("Light state is undeclared");
        }
    }

private:
    std::array<int, 4> ip;
    int port;
    std::string lightState;
    double brightnessLevel;

    bool isValidIP(const std::array<int, 4>& ip) const {
        for (const auto& octet : ip) {
            if (!(0 <= octet && octet <= 255)) {
                return false;
            }
        }
        return true;
    }

    bool isValidPort(int port) const {
        return 0 <= port && port <= 65535;
    }

    bool isValidBrightness(double brightness) const {
        return 0 <= brightness && brightness <= 1;
    }
};



// part 3
std::string lstrip(const std::string& s, char characterToStrip) {
    size_t start = 0;
    while (start < s.size() && s[start] == characterToStrip) {
        start++;
    }
    return s.substr(start);
}
struct Dataframe {
  int sl;
  std::string id;
  std::string last;
  std::string first;
  std::string middle;
};
struct FinalDataframe {
  std::string id;
  std::string last;
  std::string first;
  std::string middle;
};

std::string student_merge(std::string fname, int* p_n_students) {
  json j;
  std::ifstream jin;
  jin.open(fname);
  jin >> j;

  int arr_size = j.size();
  int sl = 0;

  std::string id_arr[arr_size];
  std::vector<Dataframe> data;
  std::map<int, std::string> ids;

  for (int i = 0; i < arr_size; i++) {
      std::string temp_str = j[i][0];
    if(temp_str[1]== '0'){
      temp_str = temp_str.substr(1);
      temp_str = lstrip(temp_str,'0');
      temp_str = 'u' + temp_str;
    }

    if (j[i].size() == 4) {
      data.push_back({sl, temp_str, j[i][1], j[i][2], j[i][3]});
    }
    else {
      data.push_back({sl, temp_str, j[i][1], j[i][2], ""});
    }
    id_arr[i] = temp_str;
    ids[sl] = temp_str;
    sl++;
  }

  std::unordered_map<std::string, int> lastOccurrences;

    // Iterate through the original map and update the last occurrences
  for (auto it = ids.begin(); it != ids.end(); ++it) {
      lastOccurrences[it->second] = it->first;
  }

    // Rebuild the map with only the last occurrences
  std::map<int, std::string> resultMap;
  for (const auto& entry : ids) {
      if (entry.first == lastOccurrences[entry.second]) {
          resultMap[entry.first] = entry.second;
      }
  }
  // Formatting to U-xxx-yyy-zzz format
  for (auto it = resultMap.begin(); it != resultMap.end(); ++it) {
        int len = it->second.length();
        int target = 10 - len;
        std::string pad(target, '0');
        std::string fs = "U" + pad + it->second.substr(1);
        std::string id = fs.substr(0, 4) + '-' + fs.substr(4, 3) + '-' + fs.substr(7);
        it->second = id;
  }

  std::vector<FinalDataframe> df;

  //Removing duplicates and only keeping the values added last along with the rest.
  for (const auto& record : data) {
    int slValue = record.sl;
    auto it = resultMap.find(slValue);
    if (it != resultMap.end()) {
      df.push_back({resultMap[slValue], record.last, record.first, record.middle});
    }
  }

  *p_n_students = df.size();        //Final Count of the df
  json student_df;
  for(const auto& student : df) { 
    student_df.push_back({
      {"id", student.id},
      {"last", student.last},
      {"first", student.first},
      {"middle", student.middle}
    });
  }
  std::string student_database = student_df.dump();     //JSON of the df
  
  jin.close();
  return student_database;
}

File: hw6/hw6_midtermexam.py
# Copyright 2023 Yuqi Jin yuqijin8@bu.edu


# Part one
"""
Write a function with the following signature:

   def translate(msg,convert):


that converts the values of a sequence (msg) based on a translation dictionary (convert)
and returns the new sequence as a tuple.

This is best described with examples.


message = (5,6,7)
converter = {7:"this", 5:"a", "a":"one"}
a = translate(message, converter) 
print(a)


The result will be

    ('a', 6, 'this')

Basically, if an element of the message is a key of the translation dictionary, it should
be "translated" by being replaced with the value associated with that key in the translation dictionary.


Here are a few other examples

translate( (1,2,3),{1:"one",2:"two",3:"three"}) -> ("one","two","three")      # numbers to English 
translate( ("a","b","c"),{"a":"z","b":"y","c":"x"}) -> ("z","y","x")          # substitution cipher
translate( ("I","am","Paul"), {"am":"suis","I":"Je"}) -> ("Je","suis","Paul") # English to French
translate( ["I","am","Paul"], {}) -> ("I","am","Paul")                        # no translation
translate( range(5), {}) -> (0, 1, 2, 3, 4)                                   # no translation


"""

# You may include test code to test your function `translate`. 
# However, your program, if imported, should run through (no inputs, etc)  
def translate(msg, convert):
    message = []
    for char in msg:
        if char in convert:
            message.append(convert[char])
        else:
            message.append(char)
    
    return tuple(message)

# part 2
"""
Write a class 'SmartBulb' that models an internet-enabled 
dimmable light bulb:

- the constructor should require that it be provided an IPv4 address
  in the form of a 4-tuple of 8-bit integers, and a port number in int16 format.

- when the light is first created, the light is neither on nor off. It is "undeclared."

- the light can be turned off with an "off" method, "on" with a "on" method, or
dimmed to 0-100% with a "dim" method.

- objects of type SmartBulb should evaluate to "True" when the light is on, and 
"False" when the light is off.

- the brightness can be discovered with a "brightness" method-- returning a number
  between 0 (completely off) and 1 (fully on).

- whenever the requirements are not satisfied -- for the constructor, for the dim method,
or for the True/False evaluation, a ValueError exception should be raised.

Here is some example test code:

```
s = SmartBulb( (10,0,0,234), 64123 )
s.off()

if s:
  print("the light is on")

s.dim(0.3)
if s:
  print("the light is at", s.brightness()*100, "percent")

s.on()
if s:
  print("the light is at", s.brightness()*100, "percent")

q = SmartBulb( (10,0,0,234), 64123 )
try:
  if q:
    print("the light is on")
except:
  print("???")
```

The test code shown should print

the light is at 30.0 percent
the light is at 100 percent
???


"""

# You may include test code to test your class SmartBulb
# However, your program, if imported, should run through (no inputs, etc)
class SmartBulb:
    # build a constructor IPv4 address and port_number
    def __init__(self, ip_address, port):
        if not isinstance(ip_address, tuple) or len(ip_address) != 4:
            raise ValueError("The form of IPv4 address is invalid")
        if not all(0 <= octet <= 255 for octet in ip_address):
            raise ValueError("The form of IPv4 address is not 8-bit integers")
        if not isinstance(port, int):
            raise ValueError("The form of port number is invalid")
        if not 0 <= port <= 65535:
            raise ValueError("The form of port number is not int16 format")
        self.ip_address = ip_address
        self.port = port
        self.state = None # default bulb state is None also means undeclared
        self.brightness_level = 0.0  # default brightness value
    
    # build on method
    # on is true
    def on(self):
        self.state = True
        self.brightness_level = 1.0
    
    # build off method
    # off is false
    def off(self):
        self.state = False
        self.brightness_level = 0.0
    
    def dim(self, level):
        if not isinstance(level, (int, float)):
            raise ValueError("The input level value type is not valid")
        if not (0 <= level <= 100):
            raise ValueError("The brightness level should be between 0 and 100")
        self.state = True if level > 0 else False
        self.brightness_level = level 
    
    # build brightness method
    def brightness(self):
        return self.brightness_level
    
    # True/False evaluation
    # when state is None, it means undeclared
    def __bool__(self): 
        if self.state is None:
            raise ValueError("The state of the bulb is undeclared")
        return self.state

# example test case #
# s = SmartBulb( (10,0,0,234), 64123 )
# s.off()

# if s:
#   print("the light is on")

# s.dim(0.3)
# if s:
#   print("the light is at", s.brightness()*100, "percent")

# s.on()
# if s:
#   print("the light is at", s.brightness()*100, "percent")

# q = SmartBulb( (10,0,0,234), 64123 )
# try:
#   if q:
#     print("the light is on")
# except:
#   print("???")

# part 3
"""
You are provided a database of student records, (as a list of tuples)
each of which has one of these formats

    student-id-formatA, LastName, Firstname, Middlename
    student-id-formatA, LastName, Firstname

    student-id-formatB, LastName, Firstname, Middlename
    student-id-formatB, LastName, Firstname

The two formats for student IDs look like this

student-id-formatA:  uX  where x is a decimal number between 1000 and 999999
student-id-formatB:  uqqqqqqqq where `qqqqqqqq` is an eight-digit zero-filled number.

There is overlap in these formats, in the sense that if the numerical value
of `X` and `qqqqqqqq` are the same, then this is the same person.

The university wants to merge these records and use a new format called "student-id-formatC"
for all student ids, like this:

    Uxxx-yyy-zzz

The new student number `xxxyyyzzz` is a nine-digit zero-filled decimal number


Other facts:

- some students can appear in the database more than once: either with or 
without a middle name, and either in ID-A format or ID-B format.

- any two people with the same student number (whether written in A, B, or C format)  
*ARE* the same person, even if their names appear differently.

YOUR TASKS

1. Count the number of unique students in the database, naming the result  `n_students`

2. Convert the database into one with a single format: a JSON object
  which is an array of student objects, each object in the format

  {"id":formatC, "last":LastName, "first":FirstName, "middle":MiddleName}

  You should use `json.dumps` to create your final database, with name `student_database`

  NOTE NOTE NOTE
  --------------

  DO NOT create a new file: if you use json.dump instead of json.dumps you will get a zero
  on this part of the exam.


Specifications
--------------

- If the middle name is empty, then the key "middle" should still be present, but
  its value should be the empty string.

- Duplicates should be eliminated, with preference to those records that
  had a middle name included.

- If a student appears twice or more in the original database with all three names 
  defined, the record that appears later in the database (by index) should be kept. 
  This represents the situation whereby an individual has changed their name.


The database to use is available on the course website at 


 https://curl.bu.edu:9602/inclass/raw_student_info.json

as a json file.

"""
"""Write a python function student_merge with the following signature
def student_merge(fname : str ) -> int, str
    that returns the values of n_students and student_database as described in the midterm exam part three. """
import json
import pandas as pd

def convert_to_format_C(str, length):
    start_letter = str[0]
    numbers = str[1:]
    result = start_letter + numbers.zfill(length - 1)
    result = result[0:4] + '-' + result[4:7] + '-' + result[7:]
    result = result.capitalize()
    return result

def student_merge(fname: str) -> (int, str):
    with open(fname) as file:
        db = json.load(file)
        
    for record in db:
        # even they don't have middle name, append the empty string
        if len(record) == 3:
            record.append("")
        if record[0].startswith('u0'):
            # cut the leading zeros if this id has 
            record[0] = record[0][0] + record[0][1:].lstrip('0')
    
    # create new panda frame{"id":formatC, "last":LastName, "first":FirstName, "middle":MiddleName}
    df = pd.DataFrame(db, columns = ["id", "last", "first", "middle"])
    df = df.drop_duplicates(subset = ['id'], keep = 'last')
    
    df['id'] = df['id'].apply(convert_to_format_C, length = 10)
    n_students = df['id'].nunique()
    student_database = json.dumps(df.to_dict(orient='records'), indent=2)
    
    return n_students, student_database

# fname = "raw_student_info.json"
# n_students, student_database = student_merge(fname)
# print("Number of unique students:", n_students)
# print("Student Database:", student_database)